src/vdaf/hits.rs

```diff
@@ -0,0 +1,897 @@
+//! multiple rounds. Question: Will this require API changes to [`crate::vdaf::Vdaf`]?
+//!
+//! TODO Update trait [`Idpf`] so thtat the IPDF can have a different field type at the leaves than
+//! at the inner nodes.
+//!
+//! TODO Implement the efficient IDPF of [[BBCG+21]]. [`ToyIdpf`] is not space efficient and is
+//! merely intended as a proof-of-concept.
+//!
+//! [VDAF]: https://cjpatton.github.io/vdaf/draft-patton-cfrg-vdaf.html
+//! [BBCG+21]: https://eprint.iacr.org/2021/017
+
+use std::array::IntoIter;
+use std::cmp::Ordering;
+use std::collections::{BTreeMap, BTreeSet};
+use std::convert::{TryFrom, TryInto};
+use std::marker::PhantomData;
+
+use crate::field::{split_vector, FieldElement};
+use crate::fp::log2;
+use crate::prng::Prng;
+use crate::vdaf::suite::{Key, KeyDeriver, KeyStream, Suite};
+use crate::vdaf::{
+    Aggregatable, Aggregator, Client, Collector, PrepareStep, PrepareTransition, Share, Vdaf,
+    VdafError,
+};
+
+/// An input for an IDPF ([`Idpf`]).
+//
+// TODO Make this an associated type of `Ipdf`.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct IdpfInput {
+    index: usize,
+    level: usize,
+}
+
+impl IdpfInput {
+    /// Constructs an IDPF input using the first `len` bits of `data`.
+    pub fn new(data: &[u8], level: usize) -> Result<Self, VdafError> {
+        if level > data.len() << 3 {
+            return Err(VdafError::Uncategorized(format!(
+                "desired bit length ({} bits) exceeds data length ({} bytes)",
+                level,
+                data.len()
+            )));
+        }
+
+        let mut index = 0;
+        let mut i = 0;
+        for byte in data {
+            for j in 0..8 {
+                let bit = (byte >> j) & 1;
+                if i < level {
+                    index |= (bit as usize) << i;
+                }
+                i += 1;
+            }
+        }
+
+        Ok(Self { index, level })
+    }
+
+    /// Construct a new input that is a prefix of `self`. Bounds checking is performed by the
+    /// caller.
+    fn prefix(&self, level: usize) -> Self {
+        let index = self.index & ((1 << level) - 1);
+        Self { index, level }
+    }
+
+    /// Return the position of `self` in the look-up table of `ToyIdpf`.
+    fn data_index(&self) -> usize {
+        self.index | (1 << self.level)
+    }
+}
+
+impl PartialOrd for IdpfInput {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for IdpfInput {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.level.cmp(&other.level) {
+            Ordering::Equal => self.index.cmp(&other.index),
+            ord => ord,
+        }
+    }
+}
+
+/// An Incremental Distributed Point Function (IDPF), as defined by [[BBCG+21]].
+///
+/// [BBCG+21]: https://eprint.iacr.org/2021/017
+//
+// NOTE(cjpatton) The real IDPF API probably needs to be stateful.
+pub trait Idpf<const KEY_LEN: usize, const OUT_LEN: usize>: Sized {
```

Comment:
Will we ever have more than one implementation of trait `Idpf`? IIUC `ToyIdpf` will eventually be replaced by a more efficient and scalable implementation, so could we just have `pub struct Idpf<const KEY_LEN: usize, const OUT_LEN: usize, F>`?

Reply:
`ToyIdpf` SHOULD be replaced eventually, but I think we want to keep things generic here. There's likely room for improvement over the IDPF from the paper, so we should keep the door open for "dropping in" future schemes.